src/Workspaces/Core/Portable/Diagnostics/Extensions.cs

```diff
@@ -429,4 +432,76 @@ await suppressionAnalyzer.AnalyzeAsync(
+    /// <list type="bullet">
+    ///    <item>Changes to methods inside the current project will be reflected to compute updated diagnostics.
+    ///        <see cref="Project.GetDependentSemanticVersionAsync(CancellationToken)"/> does not change as it only returns top level changes.</item>
+    ///    <item>Reloading a project without making any changes will re-use cached diagnostics.
+    ///        <see cref="Project.GetDependentSemanticVersionAsync(CancellationToken)"/> changes as the project is removed, then added resulting in a version change.</item>
+    /// </list>   
+    /// </para>
+    /// This checksum is also affected by the <see cref="SourceGeneratorExecutionVersion"/> for this project.
+    /// As such, it is not usable across different sessions of a particular host.
+    /// </remarks>
+    public static Task<Checksum> GetDiagnosticChecksumAsync(this Project? project, CancellationToken cancellationToken)
+    {
+        if (project is null)
+            return SpecializedTasks.Default<Checksum>();
+
+        var lazyChecksum = s_projectToDiagnosticChecksum.GetValue(
+            project,
+            static project => AsyncLazy.Create(
+                static (project, cancellationToken) => ComputeDiagnosticChecksumAsync(project, cancellationToken),
+                project));
+
+        return lazyChecksum.GetValueAsync(cancellationToken);
+
+        static async Task<Checksum> ComputeDiagnosticChecksumAsync(Project project, CancellationToken cancellationToken)
+        {
+            var solution = project.Solution;
+
+            using var _ = ArrayBuilder<Checksum>.GetInstance(out var tempChecksumArray);
+
+            // Mix in the SG information for this project.  That way if it changes, we will have a different
+            // checksum (since semantics could have changed because of this).
+            if (solution.CompilationState.SourceGeneratorExecutionVersionMap.Map.TryGetValue(project.Id, out var executionVersion))
+                tempChecksumArray.Add(executionVersion.Checksum);
+
+            // Get the checksum for the project itself.  Note: this will normally be cached.  As such, even if we
+            // have a different Project instance (due to a change in an unrelated project), this will be fast to
+            // compute and return.
+            var projectChecksum = await project.State.GetChecksumAsync(cancellationToken).ConfigureAwait(false);
+            tempChecksumArray.Add(projectChecksum);
+
+            // Calculate a checksum this project and for each dependent project that could affect semantics for this
+            // project. We order the projects guid so that we are resilient to the underlying in-memory graph structure
+            // changing this arbitrarily.
+            foreach (var projectRef in project.ProjectReferences.OrderBy(r => r.ProjectId.Id))
```

Comment:
>ProjectReferences [](http://example.com/codeflow?start=47&length=17)

I don't think it's a problem, but this could hit repeat projects during the recursion, whereas it wouldn't have before, right?